monero-wallet/src/lib.rs

```diff
@@ -0,0 +1,93 @@
+use anyhow::{Context, Result};
+use monero::consensus::encode::VarInt;
+use monero::cryptonote::hash::Hashable;
+use monero_rpc::monerod;
+use monero_rpc::monerod::{GetBlockResponse, MonerodRpc as _};
+use rand::Rng;
+
+pub struct Wallet {
+    client: monerod::Client,
+}
+
+impl Wallet {
+    /// Chooses 10 random key offsets for use within a new confidential
+    /// transactions.
+    ///
+    /// Choosing these offsets randomly is not ideal for privacy, instead they
+    /// should be chosen in a way that mimics a real spending pattern as much as
+    /// possible.
+    pub async fn choose_ten_random_key_offsets(&self) -> Result<[VarInt; 10]> {
+        let latest_block = self.client.get_block_count().await?;
+        let latest_spendable_block = latest_block.count - 10;
+
+        let block: GetBlockResponse = self.client.get_block(latest_spendable_block).await?;
+
+        let tx_hash = block
+            .blob
+            .tx_hashes
+            .first()
+            .copied()
+            .unwrap_or_else(|| block.blob.miner_tx.hash());
+
+        let indices = self.client.get_o_indexes(tx_hash).await?;
+
+        let last_index = indices
+            .o_indexes
+            .into_iter()
+            .max()
+            .context("Expected at least one output index")?;
+        let oldest_index = last_index - (last_index / 100) * 40; // oldest index must be within last 40% TODO: CONFIRM THIS
```

Comment:
Is this a protocol constraint?

Reply:
I am not fully sure yet. It is blindly copied based on the first caveat here: https://github.com/kayabaNerve/cryptonote-library#caveats